python/python_helper_sympy.py

```diff
@@ -221,71 +325,206 @@ def evaluate(expr, locals_for_eval={}):
 
     # Convert AST to code and evaluate it with no global expressions and with
     # a whitelist of local expressions
-    locals = {}
-    for key in locals_for_eval:
-        locals = {**locals, **locals_for_eval[key]}
-    return eval(compile(root, '<ast>', 'eval'), {'__builtins__': None}, locals)
+    locals = {
+        name: expr
+        for local_expressions in locals_for_eval.values()
+        for name, expr in cast(SympyMapT, local_expressions).items()
+    }
+
+    return eval(compile(root, "<ast>", "eval"), {"__builtins__": None}, locals)
 
-def convert_string_to_sympy(a, variables, allow_hidden=False, allow_complex=False):
+
+def convert_string_to_sympy(
+    expr: str,
+    variables: Optional[List[str]],
+    *,
+    allow_hidden: bool = False,
+    allow_complex: bool = False,
+    allow_trig_functions: bool = True,
+) -> sympy.Expr:
     const = _Constants()
 
     # Create a whitelist of valid functions and variables (and a special flag
     # for numbers that are converted to sympy integers).
-    locals_for_eval = {
-        'functions': const.functions,
-        'variables': const.variables,
-        'helpers': const.helpers,
+    locals_for_eval: LocalsForEval = {
+        "functions": const.functions,
+        "variables": const.variables,
+        "helpers": const.helpers,
     }
     if allow_hidden:
-        locals_for_eval['variables'] = {**locals_for_eval['variables'], **const.hidden_variables}
+        locals_for_eval["variables"].update(const.hidden_variables)
     if allow_complex:
-        locals_for_eval['variables'] = {**locals_for_eval['variables'], **const.complex_variables}
+        locals_for_eval["variables"].update(const.complex_variables)
         if allow_hidden:
-            locals_for_eval['variables'] = {**locals_for_eval['variables'], **const.hidden_complex_variables}
+            locals_for_eval["variables"].update(const.hidden_complex_variables)
+
+    if allow_trig_functions:
+        locals_for_eval["functions"].update(const.trig_functions)
 
     # If there is a list of variables, add each one to the whitelist
     if variables is not None:
-        for variable in variables:
-            locals_for_eval['variables'][variable] = sympy.Symbol(variable)
+        locals_for_eval["variables"].update(
+            (variable, sympy.Symbol(variable)) for variable in variables
+        )
 
     # Do the conversion
-    return evaluate(a, locals_for_eval)
+    return evaluate(expr, locals_for_eval)
+
 
-def point_to_error(s, ind, w = 5):
-    w_left = ind - max(0, ind-w)
-    w_right = min(ind+w, len(s)) - ind
-    return s[ind-w_left:ind+w_right] + '\n' + ' '*w_left + '^' + ' '*w_right
+def point_to_error(s: str, ind: int, w: int = 5) -> str:
```

Comment:
I'm not super happy with the readability of this function -- what does `w` mean, and where do all the shifts come from?

Reply:
This is one of those kinda magic functions that was in this file for controlling something in the display. I'll add a comment about what this does, but something I don't really want to alter _too_ much.